### PR TITLE
[move] Implemented detection of Move version mismatch

### PR DESCRIPTION
**Pull request by awelc**
This resolves https://github.com/MystenLabs/sui/issues/3471 describing a problem developers were hitting when building their packages with a binary whose version was incompatible with the version of the framework used as a package dependency (e.g., building with a binary built off devnet branch but with the framework version in the dependency corresponding to the main branch). The error that was being reported in case incompatibility was causing problems was really opaque and now if the version mismatch happens (when building or testing locally), they will see:
```
~/sui/target/debug/sui move test
Failed to verify the Move module, reason: "Sui framework version mismatch between package dependencies and Sui binary".
```

The idea is to compare `CompiledModule`s for a version that is "bundled" with the binary (we already had this data available but a bit of refactoring was required to make it accessible here) and a version that comes as a package dependency.

### CODE REVIEW

File: crates/sui-framework/src/lib.rs [hunk 1]
```diff
@@ -132,6 +137,58 @@ pub fn run_move_unit_tests(
...
+        .iter()
+        .filter(|m| *m.self_id().address() == SUI_FRAMEWORK_ADDRESS)
+        .collect();
+
+    let framework_modules = Modules::new(get_sui_framework().iter()).iter_modules_owned();
+    let framework: Vec<&CompiledModule> = framework_modules.iter().collect();
+
+    if dep_framework != framework {
+        return Err(SuiError::ModuleVerificationFailure {
+            error: "Sui framework version mismatch between package dependencies and Sui binary"
```
Comment (tnowacki) [inline crates/sui-framework/src/lib.rs:167]: We should give a suggestion of what to do here. What would the standard thing be to do? Just say "make sure your `sui` command line tool was built from the same branch/git checkout you are using for your Sui Framework files" (something like that)

File: crates/sui-framework/src/lib.rs [hunk 2]
```diff
@@ -132,6 +137,58 @@ pub fn run_move_unit_tests(
...
+    // We stash compiled modules in the Modules map which is sorted so that we can compare sets of
+    // compiled modules directly.
+
+    let dep_framework_modules = pkg.all_modules_map().iter_modules_owned();
+    let dep_framework: Vec<&CompiledModule> = dep_framework_modules
+        .iter()
+        .filter(|m| *m.self_id().address() == SUI_FRAMEWORK_ADDRESS)
+        .collect();
+
+    let framework_modules = Modules::new(get_sui_framework().iter()).iter_modules_owned();
```
Comment (tnowacki) [inline crates/sui-framework/src/lib.rs:162]: What does `get_sui_framework` do?

**Review by tnowacki — COMMENTED** [review]
Hooray! Just a few comments and then should be good to go, thanks!

File: crates/sui-framework/src/lib.rs (hunk 2 above)
Comment (awelc) [reply crates/sui-framework/src/lib.rs:162]: It forces lazy evaluation of a static:
https://github.com/MystenLabs/sui/blob/3c6e934e387c17ad86d6d0793773a7cb05ff041d/crates/sui-framework/src/lib.rs#L46

This static grabs a serialized version of the framework packages crated in the pre-build [script](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/build.rs) and deserializes it to a vector of `CompiledModules`.

File: crates/sui-framework/src/lib.rs [hunk 3]
```diff
@@ -164,7 +164,8 @@ fn verify_framework_version(pkg: &CompiledPackage) -> SuiResult<()> {
 
     if dep_framework != framework {
         return Err(SuiError::ModuleVerificationFailure {
-            error: "Sui framework version mismatch between package dependencies and Sui binary"
+            error: "Sui framework version mismatch detected.\
+                    Make sure that the sui command line tool and the Sui framework code used as a dependency correspond to the same git commit"
```
Comment (tnowacki) [inline crates/sui-framework/src/lib.rs:168]: Does this fit on one line (in Rust)? if not a line break would be nice, just rendered a bit weirdly on my screen currently (but might just be a device issue)

**Review by tnowacki — APPROVED** [review]

File: crates/sui-framework/src/lib.rs (hunk 3 above)
Comment (awelc) [reply crates/sui-framework/src/lib.rs]: Yeah, it is a little wide, but I was not sure if it's better to leave it or try to break it as I was not sure what the most natural way would be. Here's one try.

**Pull request merged**